repository-azure/src/main/java/io/aiven/elasticsearch/repositories/azure/AzureRepositoryStorageIOProvider.java

```diff
@@ -0,0 +1,217 @@
+
+    static final Setting<String> CONTAINER_NAME = Setting.simpleString("container_name");
+
+    public AzureRepositoryStorageIOProvider(final AzureClient azureClient,
+                                            final EncryptionKeyProvider encryptionKeyProvider) {
+        super(azureClient, encryptionKeyProvider);
+    }
+
+    @Override
+    protected StorageIO createStorageIOFor(final Settings repositorySettings,
+                                           final CryptoIOProvider cryptoIOProvider) {
+        checkSettings(AzureRepositoryPlugin.REPOSITORY_TYPE, CONTAINER_NAME, repositorySettings);
+        final var containerName = CONTAINER_NAME.get(repositorySettings);
+        try {
+            return Permissions.doPrivileged(() -> new AzureStorageIO(containerName, cryptoIOProvider));
+        } catch (final IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    @Override
+    public void close() throws IOException {
+        if (Objects.nonNull(client)) {
+            client.close();
+        }
+    }
+
+    private class AzureStorageIO implements StorageIO {
+
+        private final CryptoIOProvider cryptoIOProvider;
+
+        private final BlobContainerClient blobContainerClient;
+
+        AzureStorageIO(final String containerName, final CryptoIOProvider cryptoIOProvider) {
+            this.cryptoIOProvider = cryptoIOProvider;
+            this.blobContainerClient = client.blobServiceClient().getBlobContainerClient(containerName);
+        }
+
+        @Override
+        public boolean exists(final String blobName) throws IOException {
+            try {
+                return blobContainerClient.getBlobClient(blobName).exists();
+            } catch (final Exception e) {
+                throw new BlobStoreException("Failed to check if blob [" + blobName + "] exists", e);
+            }
+        }
+
+        @Override
+        public InputStream read(final String blobName) throws IOException {
+            return Permissions.doPrivileged(() ->
+                    cryptoIOProvider.decryptAndDecompress(
+                            blobContainerClient.getBlobClient(blobName).openInputStream()));
+        }
+
+        @Override
+        public void write(final String blobName,
+                          final InputStream inputStream,
+                          final long blobSize,
+                          final boolean failIfAlreadyExists) throws IOException {
+            try {
+                Permissions.doPrivileged(() -> {
+                    final var azureOutputStream =
+                            blobContainerClient
+                                    .getBlobClient(blobName)
+                                    .getBlockBlobClient()
+                                    .getBlobOutputStream(true); //always overwrite
+                    cryptoIOProvider.compressAndEncrypt(
+                            inputStream,
+                            new OutputStream() {
```

Comment:
One More thing. TBH this class worried be a bit. It was not trivial try to analyse, if it has a memory leak or not. Could we just move it to a separate class/file?

Reply:
it is not a problem, since `compressAndEncrypt` method closes all stream inside it. the ES "rule" in plugins is this (keep it in mind):
1. All `OutputStream` must be closed in plugin
2. All `InputStream` will be closed by ES snapshot service